roofit/roofitcore/inc/RooAbsCategory.h

```diff
@@ -86,38 +97,101 @@ class RooAbsCategory : public RooAbsArg {
 
   RooAbsArg *createFundamental(const char* newname=0) const;
 
-  std::vector<RooCatType*>::const_iterator begin() const {
-    return _types.cbegin();
+  /// Iterator for category state names. Points to pairs of index and name.
+  std::map<std::string, value_type>::const_iterator begin() const {
+    return stateNames().cbegin();
+  }
+  /// Iterator for category state names. Points to pairs of index and name.
+  std::map<std::string, value_type>::const_iterator end() const {
+    return stateNames().cend();
+  }
+  /// Number of states defined.
+  std::size_t size() const {
+    return stateNames().size();
   }
 
-  std::vector<RooCatType*>::const_iterator end() const {
-    return _types.cend();
+  /// Access the map of state names to index numbers. Triggers a recomputation
+  /// if the shape is dirty.
+  const std::map<std::string, value_type>& stateNames() const {
+    return const_cast<RooAbsCategory*>(this)->stateNames();
+  }
```

Comment:
this `const_cast` makes the `const` method thread-unsafe. if it's possible that someone might, in the future, care about thread-safety of `RooAbsCategory`, consider marking whatever state `clearShapeDirty` modifies `mutable` and protect accesses to it with a lock

Reply:
That's true. Thread safety is not an issue yet, but surely, I shouldn't make things worse ...